Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,223 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// Errors which may be encountered when running argument parser.
+///
+/// - unknown:            An unknown option is encountered.
+/// - expectedValue:      Expected a value from the option.
+/// - unexpectedArgument: An unexpected positional argument encountered.
+/// - expectedArguments:  Expected these positional arguments but not found.
+/// - typeMismatch:       The type of option's value doesn't match.
+public enum ArgumentParserError: Swift.Error {
+    case unknown(option: String)
+    case expectedValue(option: String)
+    case unexpectedArgument(String)
+    case expectedArguments([String])
+    case typeMismatch(String)
+}
+
+/// An abstract class representing positional or options argument.
+// FIXME: Consider renaming to Argument once we get rid of OptionsParser.
+public class CommandLineArgument: CustomStringConvertible, Hashable {
+    /// The type for the value of the argument.
+    public enum Kind {
+        case string
+        case bool
+        case int
+        case stringArray
+    }
+
+    /// The name of the argument.
+    fileprivate let name: String
+
+    /// The kind of the argument
+    fileprivate let kind: Kind
+
+    /// The usage text associated with this argument.
+    fileprivate let usage: String?
+
+    fileprivate init(name: String, kind: Kind, usage: String?) {
+        self.name = name
+        self.kind = kind
+        self.usage = usage
+    }
+
+    // MARK:- Protocol conformances
+
+    public var description: String {
+        return name + "(\(kind))"
+    }
+
+    public var hashValue: Int {
+        return name.hashValue
+    }
+
+    public static func ==(_ lhs: CommandLineArgument, _ rhs: CommandLineArgument) -> Bool {
+        return lhs.name == rhs.name
+    }
+}
+
+/// A class representing option arguments. These are optional arguments
+/// which may or may not be provided in the command line. They are always
+/// prefixed by their name. For e.g. --verbosity true.
+public class OptionArgument: CommandLineArgument {
+    override init(name: String, kind: Kind, usage: String?) {
+        precondition(!name.isPositionalArg)
```

Comment:
I'm not a big fan of using a String extension for this, I find it confusing -- I think a static method or free function would be more appropriate